### PR TITLE
feat(rpc): create wrappers for jsonrpsee clients

### LINKED ISSUE #100: rpc: re-export jsonrpsee HttpClient and WsClient
`celestia-rpc` provides convenience functions to return a `HttpClient` and `WsClient`, but no way to actually name them other than importing `jsonrpsee` directly.

celestia-rpc should reexport these (and potentially all of `jsonrpsee_http_client`, `jsonrpsee_ws_client` and maybe others, haven't checked) so that downstream users need not directly depend on celestia-rpc's version of jsonrpsee.

### PR DESCRIPTION
**Pull request by zvolin**
An alternative way to fix #100 that avoids re-exporting jsonrpsee directly

### CODE REVIEW

**Comment by oblique** [comment]
In this case, IMO we should have only `Client` and internally choose which one to use based on the uri.

File: rpc/src/client.rs [hunk 1]
```diff
@@ -1,45 +1,150 @@
...
+            if let Some(token) = auth_token {
+                let val = HeaderValue::from_str(&format!("Bearer {token}"))?;
+                headers.insert(header::AUTHORIZATION, val);
+            }
 
-        let client = HttpClientBuilder::default()
-            .set_headers(headers)
-            .build(conn_str)?;
+            let client = match conn_str.split_once(':') {
+                Some(("http", _)) => Client::Http(
```
Comment (oblique) [inline rpc/src/client.rs:47]: add `"https"`

File: rpc/src/client.rs [hunk 2]
```diff
@@ -1,45 +1,150 @@
...
+                    HttpClientBuilder::default()
+                        .set_headers(headers)
+                        .build(conn_str)?,
+                ),
+                Some(("ws", _)) => Client::Ws(
```
Comment (oblique) [inline rpc/src/client.rs:52]: add `"wss"`

File: rpc/tests/utils/client.rs [hunk 3]
```diff
@@ -54,12 +53,13 @@ pub async fn new_test_client(auth_level: AuthLevel) -> Result<WsClient> {
 }
 
 // This can be used if you want to inspect the requests from `mitmproxy`.
-pub async fn new_test_client_http(auth_level: AuthLevel) -> Result<HttpClient> {
+pub async fn new_test_client_http(auth_level: AuthLevel) -> Result<Client> {
```
Comment (oblique) [inline rpc/tests/utils/client.rs:56]: Actually we can remove this now. `Client` will take the decision based on `CELESTIA_RPC_URL` env var.

File: rpc/src/client.rs [hunk 4]
```diff
@@ -1,45 +1,150 @@
...
 
-        Ok(client)
+            Ok(client)
+        }
     }
 
-    pub async fn new_websocket(conn_str: &str, auth_token: Option<&str>) -> Result<WsClient> {
-        let mut headers = HeaderMap::new();
+    #[async_trait]
+    impl ClientT for Client {
```
Comment (oblique) [inline rpc/src/client.rs:66]: I thought we wanted to hide anything related to `jsonrpsee`. Am I missing something?

File: rpc/src/client.rs (hunk 1 above)
Comment (zvolin) [reply rpc/src/client.rs:47]: https and ws are not supported by celestia-node currently

File: rpc/src/client.rs (hunk 2 above)
Comment (zvolin) [reply rpc/src/client.rs:52]: as with https

File: rpc/src/client.rs (hunk 4 above)
Comment (zvolin) [reply rpc/src/client.rs:66]: yes, so the #100 was about having to manually include `jsonrpsee` in dependencies for typing the client we returned from `celestia_rpc::client::` helpers. We still want to have all the `(Blob|Share|Header|P2p|etc)Client` in our API and allow for interacting with them through custom `jsonrpsee` clients I think.

If we were to hide the `jsonrpsee` completely then we'd need to write a boilerplate for all the rpc modules in a `Client` and also provide more flexible interface allowing eg. setting max_request / response sizes or timeouts?

File: rpc/src/client.rs (hunk 1 above)
Comment (oblique) [reply rpc/src/client.rs:47]: Someone can use reverse proxy and have an https in front of it. I think we should support it

File: rpc/src/client.rs (hunk 4 above)
Comment (oblique) [reply rpc/src/client.rs:66]: Ok got it.

File: rpc/src/client.rs (hunk 1 above)
Comment (zvolin) [reply rpc/src/client.rs]: added

File: rpc/src/client.rs (hunk 2 above)
Comment (zvolin) [reply rpc/src/client.rs]: added

**Review by oblique — APPROVED** [review]
lgtm

**Review by fl0rek — APPROVED** [review]
nice!

**Pull request merged**